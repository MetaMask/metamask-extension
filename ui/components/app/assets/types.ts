import { Hex } from '@metamask/utils';

export type Token = {
  address: Hex;
  aggregators: string[];
  chainId: Hex;
  decimals: number;
  isNative: boolean;
  symbol: string;
  image: string;
};

export type TokenWithFiatAmount = Token & {
  tokenFiatAmount: number | null;
  balance?: string;
  string: string; // needed for backwards compatability TODO: fix this
};

export type AddressBalanceMapping = Record<Hex, Record<Hex, Hex>>;
export type ChainAddressMarketData = Record<
  Hex,
  Record<Hex, Record<string, string | number>>
>;

export type SymbolCurrencyRateMapping = Record<string, Record<string, number>>;
