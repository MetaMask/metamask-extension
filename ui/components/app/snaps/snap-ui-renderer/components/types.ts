import { ChangeEvent as ReactChangeEvent } from 'react';
import { JSXElement, SnapsChildren } from '@metamask/snaps-sdk/jsx';

export type UIComponentParams<T extends JSXElement> = {
  map: Record<string, number>;
  element: T;
  form?: string;
  useFooter?: boolean;
  onCancel?: () => void;
  promptLegacyProps?: {
    onInputChange: (event: ReactChangeEvent<HTMLInputElement>) => void;
    inputValue: string;
    placeholder?: string;
  };
  t: (key: string) => string;
  contentBackgroundColor: string | undefined;
};

export type UIComponent = {
  element: string;
  props?: Record<string, unknown>;
  children?: SnapsChildren<UIComponent | string>;
  key?: string;
};

export type UIComponentFactory<T extends JSXElement> = (
  params: UIComponentParams<T>,
) => UIComponent;
