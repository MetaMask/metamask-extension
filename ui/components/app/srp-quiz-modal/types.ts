import { ButtonSize, ButtonVariant } from '../../component-library';

export enum QuizStage {
  introduction = 'introduction',
  questionOne = 'question_one',
  wrongAnswerQuestionOne = 'wrong_answer_question_one',
  rightAnswerQuestionOne = 'right_answer_question_one',
  questionTwo = 'question_two',
  wrongAnswerQuestionTwo = 'wrong_answer_question_two',
  rightAnswerQuestionTwo = 'right_answer_question_two',
}

export type IQuizInformationProps = {
  /**
   * The icon to display in the modal should use <Icon /> component
   */
  // TODO: Replace `any` with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  icon?: any;
  /**
   * The image to display in the modal
   */
  image?: string;
  /**
   * The text content to go inside of the <Text /> component
   */
  content: string;
  /**
   * More text content to go inside of the <Text /> component
   */
  moreContent?: string;
  /**
   * Array of <Button /> component props
   */
  buttons: {
    onClick: () => void;
    label: string;
    variant: ButtonVariant;
    size?: ButtonSize;
    'data-testid'?: string;
  }[];
};

export type JSXDict = { [key: string]: () => JSX.Element };
