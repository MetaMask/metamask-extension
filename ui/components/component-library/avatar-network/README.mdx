import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import { AvatarNetwork } from './avatar-network';

# AvatarNetwork

The `AvatarNetwork` is a component responsible for display of the image of a given network

[MetaMask Design System Guides](https://www.notion.so/MetaMask-Design-System-Guides-Design-f86ecc914d6b4eb6873a122b83c12940)

<Canvas>
  <Story id="components-componentlibrary-avatarnetwork--default-story" />
</Canvas>

## Props

<ArgsTable of={AvatarNetwork} />

### Size

Use the `size` prop to set the size of the `AvatarNetwork`.

Possible sizes include:

- `xs` 16px
- `sm` 24px
- `md` 32px
- `lg` 40px
- `xl` 48px

Defaults to `md`

The fallback display of the `AvatarNetwork` is a circle with the initial letter of the network name. The size of the initial letter is calculated based on the size of the `AvatarNetwork` component.

<Canvas>
  <Story id="components-componentlibrary-avatarnetwork--size-story" />
</Canvas>

```jsx
import { AvatarNetwork, AvatarNetworkSize } from '../../component-library';

<AvatarNetwork size={AvatarNetworkSize.Xs} />
<AvatarNetwork size={AvatarNetworkSize.Sm} />
<AvatarNetwork size={AvatarNetworkSize.Md} />
<AvatarNetwork size={AvatarNetworkSize.Lg} />
<AvatarNetwork size={AvatarNetworkSize.Xl} />
```

### Name

Use the `name` prop to set the initial letter of the `AvatarNetwork`. This will be used as the fallback display if no image url is passed to the `src` prop.

Use `Text` component props to change the `variant`, `font-weight`, `font-family`, etc.

<Canvas>
  <Story id="components-componentlibrary-avatarnetwork--name" />
</Canvas>

```jsx
import { AvatarNetwork } from '../../component-library';
<AvatarNetwork name="Arbitrum One" />;
```

### Src

Use the `src` prop to set the image to be rendered of the `AvatarNetwork`.

<Canvas>
  <Story id="components-componentlibrary-avatarnetwork--src" />
</Canvas>

```jsx
import { AvatarNetwork } from '../../component-library';
<AvatarNetwork src="./images/pol-token.svg" />
<AvatarNetwork src="./images/arbitrum.svg" />
<AvatarNetwork src="./images/optimism.svg" />
<AvatarNetwork src="./images/avax-token.svg" />
<AvatarNetwork src="./images/palm.svg" />
<AvatarNetwork src="./images/bsc-filled.svg" />
<AvatarNetwork src="./images/fantom-opera.svg" />
<AvatarNetwork src="./images/harmony-one.svg" />
```

### Show Halo

Use the `showHalo` prop to display the component with halo effect. Only works if an image url is supplied to `src`

<Canvas>
  <Story id="components-componentlibrary-avatarnetwork--show-halo" />
</Canvas>

```jsx
import { AvatarNetwork } from '../../component-library';
<AvatarNetwork src="./images/arbitrum.svg" showHalo />;
```

### Color, Background Color And Border Color

Use the `color`, `backgroundColor` and `borderColor` props to set the text color, background-color and border-color of the `AvatarNetwork`.

<Canvas>
  <Story id="components-componentlibrary-avatarnetwork--color-background-color-and-border-color" />
</Canvas>

```jsx
import { TextColor,
           BackgroundColor,
           BorderColor } from '../../../helpers/constants/design-system';
import { AvatarNetwork } from '../../component-library';
<AvatarNetwork
      backgroundColor={BackgroundColor.goerli}
      borderColor={BorderColor.goerli}
      name="G"
      color={Color.goerliInverse}
>
  G
</AvatarNetwork>
<AvatarNetwork
   backgroundColor={BackgroundColor.sepolia}
      borderColor={BorderColor.sepolia}
      name="S"
      color={Color.goerliInverse}
>
  S
</AvatarNetwork>
```
