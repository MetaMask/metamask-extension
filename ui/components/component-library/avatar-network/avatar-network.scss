.mm-avatar-network {
  &--with-halo {
    position: relative;
  }

  &__network-image {
    width: 100%;

    &--blurred {
      filter: blur(8px);
      opacity: 0.5;
    }

    &--size-reduced {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      width: 62.5%;
      height: 62.5%;
      border-radius: 50%;
    }
  }
}
