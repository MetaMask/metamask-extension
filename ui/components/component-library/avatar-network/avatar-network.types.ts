import type { PolymorphicComponentPropWithRef } from '../box';
import type { AvatarBaseStyleUtilityProps } from '../avatar-base/avatar-base.types';

export enum AvatarNetworkSize {
  Xs = 'xs',
  Sm = 'sm',
  Md = 'md',
  Lg = 'lg',
  Xl = 'xl',
}

/**
 * Props for the AvatarNetwork component
 */
// TODO: Convert to a `type` in a future major version.
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export interface AvatarNetworkStyleUtilityProps
  extends Omit<AvatarBaseStyleUtilityProps, 'size' | 'children'> {
  /**
   * The name accepts the string to render the first alphabet of the Avatar Name
   */
  name: string;
  /**
   * The src accepts the string of the image to be rendered
   */
  src?: string;
  /**
   * The showHalo accepts a boolean prop to render the image with halo effect
   */
  showHalo?: boolean;
  /**
   * The size of the AvatarNetwork
   * Possible values could be AvatarNetworkSize.Xs(16px), AvatarNetworkSize.Sm(24px), AvatarNetworkSize.Md(32px), AvatarNetworkSize.Lg(40px), AvatarNetworkSize.Xl(48px)
   * Defaults to AvatarNetworkSize.Md
   */
  size?: AvatarNetworkSize;
}

export type AvatarNetworkProps<C extends React.ElementType> =
  PolymorphicComponentPropWithRef<C, AvatarNetworkStyleUtilityProps>;

export type AvatarNetworkComponent = <C extends React.ElementType = 'span'>(
  props: AvatarNetworkProps<C>,
) => React.ReactElement | null;
