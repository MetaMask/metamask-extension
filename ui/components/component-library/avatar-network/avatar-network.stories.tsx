import React from 'react';
import { StoryFn, Meta } from '@storybook/react';
import {
  Display,
  TextColor,
  BackgroundColor,
  BorderColor,
  AlignItems,
} from '../../../helpers/constants/design-system';
import { Box } from '..';
import { AvatarNetworkSize } from './avatar-network.types';
import README from './README.mdx';

import { AvatarNetwork } from './avatar-network';

export default {
  title: 'Components/ComponentLibrary/AvatarNetwork',
  component: AvatarNetwork,
  parameters: {
    docs: {
      page: README,
    },
  },
  argTypes: {
    size: {
      control: 'select',
      options: Object.values(AvatarNetworkSize),
    },
    color: {
      options: Object.values(TextColor),
      control: 'select',
    },
    backgroundColor: {
      options: Object.values(BackgroundColor),
      control: 'select',
    },
    borderColor: {
      options: Object.values(BorderColor),
      control: 'select',
    },
    name: {
      control: 'text',
    },
    src: {
      control: 'text',
    },
    showHalo: {
      control: 'boolean',
    },
  },
  args: {
    name: 'Arbitrum One',
    src: './images/arbitrum.svg',
    size: AvatarNetworkSize.Md,
    showHalo: false,
  },
} as Meta<typeof AvatarNetwork>;

const Template: StoryFn<typeof AvatarNetwork> = (args) => {
  return <AvatarNetwork {...args} />;
};

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';

export const SizeStory: StoryFn<typeof AvatarNetwork> = (args) => (
  <>
    <Box
      display={Display.Flex}
      alignItems={AlignItems.flexEnd}
      gap={1}
      marginBottom={4}
    >
      <AvatarNetwork {...args} size={AvatarNetworkSize.Xs} />
      <AvatarNetwork {...args} size={AvatarNetworkSize.Sm} />
      <AvatarNetwork {...args} size={AvatarNetworkSize.Md} />
      <AvatarNetwork {...args} size={AvatarNetworkSize.Lg} />
      <AvatarNetwork {...args} size={AvatarNetworkSize.Xl} />
    </Box>
    <Box display={Display.Flex} alignItems={AlignItems.flexEnd} gap={1}>
      <AvatarNetwork {...args} src="" size={AvatarNetworkSize.Xs} />
      <AvatarNetwork {...args} src="" size={AvatarNetworkSize.Sm} />
      <AvatarNetwork {...args} src="" size={AvatarNetworkSize.Md} />
      <AvatarNetwork {...args} src="" size={AvatarNetworkSize.Lg} />
      <AvatarNetwork {...args} src="" size={AvatarNetworkSize.Xl} />
    </Box>
  </>
);
SizeStory.storyName = 'Size';

export const Name = Template.bind({});
Name.args = {
  src: '',
};

export const Src: StoryFn<typeof AvatarNetwork> = (args) => (
  <Box display={Display.Flex} gap={1}>
    <AvatarNetwork {...args} src="./images/pol-token.svg" />
    <AvatarNetwork {...args} src="./images/arbitrum.svg" />
    <AvatarNetwork {...args} src="./images/optimism.svg" />
    <AvatarNetwork {...args} src="./images/avax-token.svg" />
    <AvatarNetwork {...args} src="./images/palm.svg" />
    <AvatarNetwork {...args} src="./images/bsc-filled.svg" />
    <AvatarNetwork {...args} src="./images/fantom-opera.svg" />
    <AvatarNetwork {...args} src="./images/harmony-one.svg" />
  </Box>
);

export const ShowHalo = Template.bind({});
ShowHalo.args = {
  showHalo: true,
};

export const ColorBackgroundColorAndBorderColor: StoryFn<
  typeof AvatarNetwork
> = (args) => (
  <Box display={Display.Flex} gap={1}>
    <AvatarNetwork
      {...args}
      backgroundColor={BackgroundColor.goerli}
      borderColor={BorderColor.goerli}
      name="G"
      color={TextColor.goerliInverse}
    />
    <AvatarNetwork
      {...args}
      backgroundColor={BackgroundColor.sepolia}
      borderColor={BorderColor.sepolia}
      name="S"
      color={TextColor.goerliInverse}
    />
  </Box>
);
ColorBackgroundColorAndBorderColor.args = {
  src: '',
};
