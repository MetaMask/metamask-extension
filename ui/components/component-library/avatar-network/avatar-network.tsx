import React, { useState, useEffect } from 'react';
import classnames from 'classnames';
import {
  Display,
  AlignItems,
  JustifyContent,
  TextColor,
  BackgroundColor,
  BorderColor,
  BorderRadius,
} from '../../../helpers/constants/design-system';
import type { PolymorphicRef } from '../box';
import { AvatarBase, AvatarBaseProps } from '../avatar-base';
import type { AvatarNetworkComponent } from './avatar-network.types';
import { AvatarNetworkProps, AvatarNetworkSize } from './avatar-network.types';

export const AvatarNetwork: AvatarNetworkComponent = React.forwardRef(
  <C extends React.ElementType = 'div'>(
    {
      size = AvatarNetworkSize.Md,
      name,
      src,
      showHalo,
      color = TextColor.textDefault,
      backgroundColor = BackgroundColor.backgroundAlternative,
      borderColor = BorderColor.backgroundDefault,
      className = '',
      ...props
    }: AvatarNetworkProps<C>,
    ref?: PolymorphicRef<C>,
  ) => {
    const [showFallback, setShowFallback] = useState(false);

    useEffect(() => {
      setShowFallback(!src);
    }, [src]);

    const fallbackString = name?.[0] ?? '?';

    const handleOnError = () => {
      setShowFallback(true);
    };

    return (
      <AvatarBase
        ref={ref}
        size={size}
        display={Display.Flex}
        alignItems={AlignItems.center}
        justifyContent={JustifyContent.center}
        className={classnames(
          'mm-avatar-network',
          showHalo ? 'mm-avatar-network--with-halo' : '',
          className,
        )}
        borderRadius={BorderRadius.MD}
        borderWidth={1}
        {...{
          backgroundColor,
          borderColor,
          color,
          ...(props as AvatarBaseProps<C>),
        }}
      >
        {showFallback ? (
          fallbackString
        ) : (
          <>
            {showHalo && (
              <img
                src={src}
                className={
                  showHalo ? 'mm-avatar-network__network-image--blurred' : ''
                }
                aria-hidden="true"
              />
            )}
            <img
              className={
                showHalo
                  ? 'mm-avatar-network__network-image--size-reduced'
                  : 'mm-avatar-network__network-image'
              }
              onError={handleOnError}
              src={src}
              alt={(name && `${name} logo`) || 'network logo'}
            />
          </>
        )}
      </AvatarBase>
    );
  },
);
