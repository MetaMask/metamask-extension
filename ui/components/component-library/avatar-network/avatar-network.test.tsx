/* eslint-disable jest/require-top-level-describe */
import { render, screen } from '@testing-library/react';
import React from 'react';

import {
  BackgroundColor,
  BorderColor,
  TextColor,
} from '../../../helpers/constants/design-system';
import { AvatarNetwork } from './avatar-network';
import { AvatarNetworkSize } from './avatar-network.types';

describe('AvatarNetwork', () => {
  const args = {
    name: 'ethereum',
    src: './images/eth_logo.svg',
    showHalo: false,
  };

  it('should render correctly', () => {
    const { getByTestId, container } = render(
      <AvatarNetwork {...args} data-testid="avatar-network" />,
    );
    expect(getByTestId('avatar-network')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('should render image of Avatar Network', () => {
    render(<AvatarNetwork data-testid="avatar-network" {...args} />);
    const image = screen.getByRole('img');
    expect(image).toBeDefined();
    expect(image).toHaveAttribute('src', args.src);
  });

  it('should render the first letter of the name prop if no src is provided', () => {
    const { getByText } = render(
      <AvatarNetwork data-testid="avatar-network" {...args} src="" />,
    );
    expect(getByText('e')).toBeDefined();
  });

  it('should render halo effect if showHalo is true and image url is there', () => {
    render(<AvatarNetwork data-testid="avatar-network" {...args} showHalo />);
    const image = screen.getAllByRole('img', { hidden: true });
    expect(image[1]).toHaveClass(
      'mm-avatar-network__network-image--size-reduced',
    );
  });

  it('should render the first letter of the name prop when showHalo is true and no image url is provided', () => {
    const { getByText } = render(
      <AvatarNetwork {...args} src="" data-testid="avatar-network" showHalo />,
    );
    expect(getByText('e')).toBeDefined();
  });
  // className
  it('should render with custom className', () => {
    const { getByTestId } = render(
      <AvatarNetwork
        {...args}
        data-testid="avatar-network"
        className="test-class"
      />,
    );
    expect(getByTestId('avatar-network')).toHaveClass('test-class');
  });
  // color
  it('should render with different colors', () => {
    const { getByTestId } = render(
      <>
        <AvatarNetwork
          {...args}
          color={TextColor.successDefault}
          data-testid={TextColor.successDefault}
        />
        <AvatarNetwork
          {...args}
          color={TextColor.errorDefault}
          data-testid={TextColor.errorDefault}
        />
      </>,
    );
    expect(getByTestId(TextColor.successDefault)).toHaveClass(
      `mm-box--color-${TextColor.successDefault}`,
    );
    expect(getByTestId(TextColor.errorDefault)).toHaveClass(
      `mm-box--color-${TextColor.errorDefault}`,
    );
  });
  // background color
  it('should render with different background colors', () => {
    const { getByTestId } = render(
      <>
        <AvatarNetwork
          {...args}
          backgroundColor={BackgroundColor.successDefault}
          data-testid={BackgroundColor.successDefault}
        />
        <AvatarNetwork
          {...args}
          backgroundColor={BackgroundColor.errorDefault}
          data-testid={BackgroundColor.errorDefault}
        />
      </>,
    );
    expect(getByTestId(BackgroundColor.successDefault)).toHaveClass(
      `mm-box--background-color-${BackgroundColor.successDefault}`,
    );
    expect(getByTestId(BackgroundColor.errorDefault)).toHaveClass(
      `mm-box--background-color-${BackgroundColor.errorDefault}`,
    );
  });
  // border color
  it('should render with different border colors', () => {
    const { getByTestId } = render(
      <>
        <AvatarNetwork
          {...args}
          borderColor={BorderColor.successDefault}
          data-testid={BorderColor.successDefault}
        />
        <AvatarNetwork
          {...args}
          borderColor={BorderColor.errorDefault}
          data-testid={BorderColor.errorDefault}
        />
      </>,
    );
    expect(getByTestId(BorderColor.successDefault)).toHaveClass(
      `mm-box--border-color-${BorderColor.successDefault}`,
    );
    expect(getByTestId(BorderColor.errorDefault)).toHaveClass(
      `mm-box--border-color-${BorderColor.errorDefault}`,
    );
  });
  // AvatarNetworkSize
  it('should render with different AvatarNetworkSize', () => {
    const { getByTestId } = render(
      <>
        <AvatarNetwork
          {...args}
          size={AvatarNetworkSize.Xs}
          data-testid={AvatarNetworkSize.Xs}
        />
        <AvatarNetwork
          {...args}
          size={AvatarNetworkSize.Sm}
          data-testid={AvatarNetworkSize.Sm}
        />
        <AvatarNetwork
          {...args}
          size={AvatarNetworkSize.Md}
          data-testid={AvatarNetworkSize.Md}
        />
        <AvatarNetwork
          {...args}
          size={AvatarNetworkSize.Lg}
          data-testid={AvatarNetworkSize.Lg}
        />
        <AvatarNetwork
          {...args}
          size={AvatarNetworkSize.Xl}
          data-testid={AvatarNetworkSize.Xl}
        />
      </>,
    );
    expect(getByTestId(AvatarNetworkSize.Xs)).toHaveClass(
      `mm-avatar-base--size-${AvatarNetworkSize.Xs}`,
    );
    expect(getByTestId(AvatarNetworkSize.Sm)).toHaveClass(
      `mm-avatar-base--size-${AvatarNetworkSize.Sm}`,
    );
    expect(getByTestId(AvatarNetworkSize.Md)).toHaveClass(
      `mm-avatar-base--size-${AvatarNetworkSize.Md}`,
    );
    expect(getByTestId(AvatarNetworkSize.Lg)).toHaveClass(
      `mm-avatar-base--size-${AvatarNetworkSize.Lg}`,
    );
    expect(getByTestId(AvatarNetworkSize.Xl)).toHaveClass(
      `mm-avatar-base--size-${AvatarNetworkSize.Xl}`,
    );
  });
  it('should forward a ref to the root html element', () => {
    const ref = React.createRef<HTMLDivElement>();
    render(<AvatarNetwork {...args} ref={ref} />);
    expect(ref.current).not.toBeNull();
    if (ref.current) {
      expect(ref.current.nodeName).toBe('DIV');
    }
  });
});
