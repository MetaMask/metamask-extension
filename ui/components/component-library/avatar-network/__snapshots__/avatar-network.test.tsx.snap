// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarNetwork should render correctly 1`] = `
<div>
  <div
    class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-network mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-md mm-box--border-color-background-default mm-box--border-width-1 box--border-style-solid"
    data-testid="avatar-network"
  >
    <img
      alt="ethereum logo"
      class="mm-avatar-network__network-image"
      src="./images/eth_logo.svg"
    />
  </div>
</div>
`;
